PARQ quantizer support for torchao's weight-only configs

lisjin opened the pull request:
This is the first step in supporting `torchao.quantize_` for PARQ trained models. I target only `Int4WeightOnlyConfig` and `IntxWeightOnlyConfig` for now since PARQ does not have activation quantization.

Instead of converting the state (e.g., scale, zero point) from PARQ's existing quantizers to torchao format, I decided to create a new quantizer `UnifTorchaoQuantizer`. This quantizer calls torchao's quantization primitives `choose_qparams_affine`, `quantize_affine`, `dequantize_affine` to ensure parity between the two QAT methods.

@metascroy It would be great if you could check the correctness of how the quantizer in `TestUnifTorchaoQuantizer.test_intx_weight_only` is initialized. I'm not sure if I missed any subtleties with int8.

metascroy commented on test/prototype/test_parq.py:
```diff
@@ -97,9 +117,96 @@ def test_ternarybit_lsbq_parq_prox(self):
+        self.compare_quantized_models(model, m_ref, quantizer, b)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
```
This should work on CPU in default QDQLayout.

Only certain layouts require CPU.

metascroy commented on test/prototype/test_parq.py:
```diff
@@ -97,9 +117,96 @@ def test_ternarybit_lsbq_parq_prox(self):
+
+        # based off torchao.quantization.quant_api._intx_weight_only_transform
+        b = 8
+        q_dtype = torch.int8
```
Why only int8?  Can we do torch.int1, torch.int2, ..., torch.int8?

lisjin replied:
Definitely! I wasn't aware these data types exist but will include them.

metascroy commented:
@lisjin can you give a little code snippet of our QAT prepare/convert would work for this API?

I'm having trouble following.  Here are some example code snippets from other APIs: https://fb.workplace.com/groups/pytorch.edge2.team/permalink/1186139489308568/

lisjin commented on test/prototype/test_parq.py:
```diff
@@ -53,52 +78,184 @@ def forward(self, x):
+        self.compare_quantized_models(model, m_ref, quantizer, b, group_size)
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    @common_utils.parametrize("group_size", [32, 512])
```
Added parametrized test cases that loop over (int2, int3, int4, int8) data types.

lisjin commented on test/prototype/test_parq.py:
```diff
@@ -53,52 +78,184 @@ def forward(self, x):
+
+        orig_model = copy.deepcopy(model)  # save copy of PARQ quantized model
+
+        # equivalent to torchao's convert step
+        with torch.no_grad():
+            optimizer.restore_latent_params()
+        quantize_(model, quantizer.config)
+
+        for n, module in model.named_modules():
+            if not _is_linear(module):
+                continue
+
+            p_orig = getattr(orig_model, n).weight  # PARQ weight
+            p = module.weight.dequantize()  # PARQ weight after quantize_
+            p_ref = getattr(m_ref, n).weight.dequantize()  # natively quantize_
+
+            self.assertTrue(p_orig.equal(p_ref))
+            self.assertTrue(p.equal(p_ref))
```
@metascroy Lines 241–243 shows how to convert a PARQ quantized model using `quantize_`. Since PARQ stores full-precision "latent" copies of each quantized parameter, I just load those values into the model before calling `quantize_`. Line 254 shows that we match the natively quantized model `m_ref`.

lisjin commented on torchao/prototype/parq/optim/quantopt.py:
```diff
@@ -106,6 +108,28 @@ def quantize_(
         return q
 
+    @torch.no_grad()
+    def torchao_quantize_(self, model):
```
@metascroy Here is a new optimizer-level method within PARQ that (1) loads the latent full-precision weights and (2) recursively quantizes all modules using `torchao.quantize_`. Just to double-check, will it also work on modules besides `torch.nn.Linear`?

metascroy replied:
Usually to do non-linear, we need a third filter_fn argument, e.g., for embedding:

```
quantize_(model, IntxWeightOnlyConfig(torch.int4), lambda m, fqn: isinstance(m, torch.nn.Embedding))
```

By default the filter function in quantize_ is torch.nn.Linear, which is why you don't usually see it specified.

andrewor14 commented:
Hi @lisjin, do you mind adding a code snippet on the main README on what the end-to-end flow would look like? My understanding is you can just replace the `LSBQuantizer` there with your new `UnifTorchaoQuantizer`. Then what happens after training? Do we call `quantize_(model, Int4WeightOnlyConfig)` as before? Would be good to clarify

andrewor14 commented on torchao/prototype/parq/quant/uniform_torchao.py:
```diff
@@ -0,0 +1,123 @@
+        eps: Optional[float] = None,
+        preserve_zero: bool = True,
+        zero_point_domain: ZeroPointDomain = ZeroPointDomain.NONE,
+        config: Optional[AOBaseConfig] = None,
```
Hmm, a bit weird that this isn't used inside the class at all. Can we just let the user pass this into `torchao_quantize_` instead?

lisjin commented:
@andrewor14 Thanks for the feedback—I removed config from UnifTorchaoQuantizer. In the README, I've also added a side-by-side comparison of PARQ vs. torchao prepare and convert steps. After PARQ training, we call `optimizer.torchao_quantize_(model, config)`. Let me know if there's anything missing.

andrewor14 commented:
Looks great, thanks @lisjin! The README is very clear.

One thing I want to discuss is whether we can just use a new `PARQConfig` instead so the PARQ flow looks more like the existing torchao QAT flow. This is current convert flow in the PR now:
```
config = IntXWeightOnlyConfig(weight_dtype=torch.int4, granularity=PerGroup(32))
optimizer.torchao_quantize_(model, config)
```

What do you think about something like this instead?
```
inner_config = IntXWeightOnlyConfig(weight_dtype=torch.int4, granularity=PerGroup(32))
parq_config = PARQConfig(optimizer, inner_config)
quantize_(model, parq_config)
```
Also curious if @metascroy has any thoughts on this

lisjin commented on torchao/prototype/parq/optim/quantopt.py:
```diff
@@ -106,6 +109,27 @@ def quantize_(
+    def torchao_quantize_(self, model: torch.nn.Module, config: AOBaseConfig):
+        """Recursively call torchao.quantize_ on model using given config."""
+        self.restore_latent_params()
+        param_set = {
+            p.data_ptr()
+            for group in self.regularized_param_groups()
+            for p in group["params"]
+        }
+
+        def inner_quantize_(model):
+            for module in model.children():
+                for param in module.parameters(recurse=False):
+                    if param.data_ptr() in param_set:
+                        quantize_(module, config)
+                        break
+
+                inner_quantize_(module)
```
@andrewor14 I think the `quantize_(model, PARQConfig(optimizer, inner_config))` interface looks cleaner, but I see a potential issue with applying this `inner_quantize_` function. PARQ uses `param_set` to filter which modules to quantize. However, `quantize_` has its own recursive function that quantizes parameters based off `filter_fn` mapping `nn.Module` -> `bool`. I'm not sure how to reconcile the two.

andrewor14 replied:
I see, yeah doesn't seem trivial to reconcile these two. Maybe this is OK then

andrewor14 replied:
Another idea is:

```
filter_fn = optimizer.get_filter_fn()
quantize_(model, config, filter_fn)
```
where `optimizer.get_filter_fn` returns a function that checks against the `param_set`. Then you don't need `PARQConfig`

andrewor14 replied:
Another idea is:

```
filter_fn = optimizer.get_filter_fn()
quantize_(model, config, filter_fn)
```
where `optimizer.get_filter_fn` returns a function that checks against the `param_set`. Then you don't need `PARQConfig`. What do you think?

andrewor14 commented on torchao/prototype/parq/optim/quantopt.py:
```diff
@@ -106,6 +109,27 @@ def quantize_(
+            for module in model.children():
+                for param in module.parameters(recurse=False):
+                    if param.data_ptr() in param_set:
+                        quantize_(module, config)
```
By the way this call itself is also recursive, so I feel this will end up quantizing the same submodules multiple times? Maybe we should express this as part of a filter_fn, something like:

```
def _filter_fn(module: torch.nn.Module) -> bool:
    for param in module.parameters(recurse=False):
        if param.data_ptr() in param_set:
            return True
    return False

quantize_(model, config, _filter_fn)
```

lisjin replied:
Good point! I'll add this change in

andrewor14 reviewed (approved):
Looks good to me other than the recursion comment. @metascroy any other thoughts?

metascroy commented:
Looks good to me!  Thanks @lisjin!

Can we add an end-to-end test_intx_weight_only_e2e for intx (with various x-values), similar to test_int4_weight_only_e2e?

metascroy commented on torchao/prototype/parq/quant/uniform_torchao.py:
```diff
@@ -0,0 +1,135 @@
+    ) -> None:
+        super().__init__(center=False)
+
+        self.mapping_type = (
```
Nit: pass mapping type directly?

metascroy commented on torchao/prototype/parq/quant/uniform_torchao.py:
```diff
@@ -0,0 +1,135 @@
+                _BIT_WIDTH_TO_DTYPE[b]
+            ]
+        if self.target_dtype is None:
+            self.target_dtype = torch.int8
```
Just make torch.int8 the default in __init__?

lisjin commented on test/prototype/test_parq.py:
```diff
@@ -53,52 +93,179 @@ def forward(self, x):
+
+    @unittest.skipIf(not TORCH_VERSION_AT_LEAST_2_6, "Test only enabled for 2.6+")
+    @unittest.skipIf(_DEVICE == "cpu", "Need GPU available")
+    @common_utils.parametrize("b", [2, 3, 4, 8])
+    def test_intx_weight_only_e2e(self, b: int = 2, group_size: int = 32):
```
@metascroy Thanks for looking it over! I've added this end-to-end test, along with `mapping_type=MappingType.SYMMETRIC` and `target_dtype=torch.int8` defaults for `UnifTorchaoQuantizer`

andrewor14 commented on test/prototype/test_parq.py:
```diff
@@ -53,52 +93,179 @@ def forward(self, x):
+
+        # equivalent to torchao's convert step
+        model.eval()
+        optimizer.restore_latent_params()
+        quantize_(model, config, filter_fn=optimizer._get_filter_fn(model))
```
Thanks! One final nit is if we expect users to call this, we should remove the underscore before `get_filter_fn` since it's now public facing API

andrewor14 commented on torchao/prototype/parq/README.md:
```diff
@@ -7,23 +7,104 @@ This library applies QAT without modifying model-level code. It instead interfac
+config = IntXWeightOnlyConfig(
+    weight_dtype=torch.int4, granularity=PerGroup(32)
+)
+optimizer.torchao_quantize_(model, config)
```
also need to update this
